guppylang/hresult.py

```diff
@@ -0,0 +1,206 @@
+r"""
+Quantinuum system results and utilities.
+
+Includes conversions to traditional distributions over bitstrings if a tagging
+convention is used, including conversion to a pytket BackendResult.
+
+Under this convention, tags are assumed to be a name of a bit register unless they fit
+the regex pattern `^([a-z][\w_]*)\[(\d+)\]$` (like `my_Reg[12]`) in which case they
+are assumed to refer to the nth element of a bit register.
+
+For results of the form ``` result("<register>", value) ``` `value` can be `{0, 1}`,
+wherein the register is assumed to be length 1, or lists over those values,
+wherein the list is taken to be the value of the entire register.
+
+For results of the form ``` result("<register>[n]", value) ``` `value` can only be
+`{0,1}`.
+The register is assumed to be at least `n+1` in size and unset
+elements are assumed to be `0`.
+
+Subsequent writes to the same register/element in the same shot will overwrite.
+
+To convert to a `BackendResult` all registers must be present in all shots, and register
+sizes cannot change between shots.
+
+"""
+
+from __future__ import annotations
+
+import re
+from collections import Counter, defaultdict
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, Literal
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+    from pytket.backends.backendresult import BackendResult
+
+#: Primitive data types that can be returned by a result
+DataPrimitive = int | float | bool
```

Comment:
I am confused by the spec description above:
> ``` result("<register>", value) ``` `value` can be `{0, 1}`
and the code here.

Are floats supported or not? I'd imagine, if guppy says `result(x, pi)`, we can return an approx value of pi as part of the results.

Or is this PR limited to bitstrings? If so, why include `float`?

Reply:
float is allowed as a result type, it is not supported under the "register bitstring convention" adopted as part of this PR. The `HResult` and `Shots` classes faithfully capture all types, it is only the post-processing/conversions that are limited.